### ISSUE TITLE
Launch context to the resource server

### ISSUE THREAD

**Issue by rsmayda** [dancinnamon-okta/okta-smartfhir-docs#3]
How will the Resource server (aka FHIR Server) know the context of the request?

The SMART app will have the context due to `patient` being next to `access_token`, but how will the resource server know?

```json
{
"access_token":"<token>",
"patient":"Patient/1234"
}
```

Use case I am concerned is how will the resource server trust the request from the SMART app?

**Comment by rsmayda** [comment]
Looks like the this launch context is added to the [claims](https://github.com/dancinnamon-okta/okta-smartfhir-docs/blob/main/SETUP.md#claims), but will the `launch_response_patient` and `fhirUser` claim also be in the `access_token`?

This approach will work for the use-case of a patient looking at their own data but will it work for a patient (Patient/parent) looking at their Childs ((Patient/child)? We would want the claims to look like:
```json
{
...
"fhirUser":"Patient/<parent>",
"launch_response_patient":"Patient/<child>",
}
```

**Comment by salpcnsi** [comment]
Here is what is in the access token, so launch_response_patient is in access token and it is coming from Okta profile of the user (user.patient_id).

{ver: 1,
  jti: 'AT.O0o-EDzPqKWzMWeZEUZbarQNIQgcKyf1B1sUtxQB340.oar4haq35KMa4I9C64x6',
  iss: 'https://dev-950xxx.okta.com/oauth2/aus1iroru5a4v27jp4x7',
  aud: 'https://esd97yyyy.execute-api.us-east-2.amazonaws.com/dev',
  iat: 1608313829,
  exp: 1608317429,
  cid: '0oa1it47noff1iY204x7',
  uid: '00udj0gep8olqMzOc4x6',
  scp: [
    'launch',
    'offline_access',
    'patient/*.*',
    'launch/patient',
    'user/*.*',
    'openid',
    'profile'
  ],
  sub: 'xxxx@gmail.com',
  launch_response_patient: '1283387'
}

**Comment by dancinnamon-okta** [comment]
A quick comment here about what I've seen (and if I'm wrong it'd be very easy to update the claims in Okta).

The fhirUser is a claim that's meant to communicate who the actual user is- regardless of patient selected.  I've found that it typically takes the format of ResourceType/ResourceID- so an example could be Patient/12345.  Now- if the user is a practitioner, the resourceType would not be "Patient", but rather something else, like "Practitioner".  So fhirUser would be Practitioner/12345.

The patient returned when the launch/patient scope is requested, can only be a patient ID- so from what i've seen in my experimentation using both the iOS SMART-FHIR SDK, and the Python SMART-FHIR SDK, the expectation is that ONLY the patient id comes back. So the format would be only "12345" there.

Also a quick note about "launch_reponse_patient" claim. Right now Okta has no out of the box way of putting extra data in the /token repsonse.  So the /token proxy is responsible for taking any "launch_response_*" claims, and putting that data in the /token response JSON.

The launch _response_patient claim can be filled in 1 of 2 ways.  The first way is as @salpcnsi mentioned- via an attribute in Okta UD.
The second way is that the token_hook is called by Okta at runtime, and the token hook will essentially take the selected patient (if a patient picker is used), and it will override whatever Okta was going to put there, if anything.  So essentially it can be done either by Okta determining the value on it's own, or via patient picker during consent.

Hope this helps!

**Comment by rsmayda** [comment]
> The fhirUser is a claim that's meant to communicate who the actual user is...

Yep, agreed! One part to note I have also seen it be an absolute URL (`fhirserver.com/Patient/12345`). This allows 1 idP to potentially support different FHIR servers (think of large hospital organization, with many separate hospitals).

> The patient returned when the launch/patient scope is requested, can only be a patient ID...

Gotcha that makes sense, so that idea here is that the Resource sever will trust the launch context within the `access_token` and it will not need to know who the `fhirUser` is.

> The launch_response_patient claim can be filled in 1 of 2 ways....

Perfect so it sounds like the Patient Selector will override the `launch_response_patient` claim and create the JWT `access_token` from there. Out of curiosity normal OAuth2 validation endpoints like /introspection & /userinfo will still work right?
